### PR TITLE
Implement Using Custom Types for Timeouts

### PR DESCRIPTION
**Pull request by bendbennett**
Closes: #17

### CODE REVIEW

File: timeouts/type/timeouts.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	return t.getTimeout(ctx, attributeNameUpdate)
+}
+
+// Delete attempts to retrieve the "delete" attribute and parse it as time.Duration.
+// If any errors are generated they are returned along with the default timeout of 20 minutes.
+func (t TimeoutsValue) Delete(ctx context.Context) (time.Duration, error) {
+	return t.getTimeout(ctx, attributeNameDelete)
+}
+
+func (t TimeoutsValue) getTimeout(_ context.Context, timeoutName string) (time.Duration, error) {
```
Comment (bflad) [inline timeouts/type/timeouts.go:106]: I would recommend switching this to `diag.Diagnostics` over `error` so its more friendly to use in provider code. 👍

File: timeouts/datasource/schema.go [hunk 2]
```diff
@@ -99,3 +112,25 @@ func attributesMap(opts Opts) map[string]schema.Attribute {
 
 	return attributes
 }
+
+func attrTypesMap(opts Opts) map[string]attr.Type {
+	attrTypes := map[string]attr.Type{}
+
+	if opts.Create {
+		attrTypes[attributeNameCreate] = types.StringType
+	}
```
Comment (bflad) [inline timeouts/datasource/schema.go:121]: Not due to this pull request, but data source schemas should only support a `read` timeout.

File: timeouts/type/timeouts.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
+package timeouts
```
Comment (bflad) [inline timeouts/type/timeouts.go:1]: I wonder if this should be promoted to the top level `timeouts` package so provider developers do not have an awkward `type` import if they do not import alias it?

**Review by bflad — COMMENTED** [review]
We should likely drop the existing `tfsdk` implementations in the same version too. 😄

**Comment by bendbennett** [comment]
> We should likely drop the existing `tfsdk` implementations in the same version too. 😄

Have dropped all usages of `tfsdk. It might be worth noting that there is no prior released version that contains deprecation notices for the `tfsdk` implementations.

**Review by bflad — COMMENTED** [review]
😅 Apparently I had this comment queued up but didn't submit it.

File: datasource/timeouts/timeouts.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+	value, ok := t.Object.Attributes()[timeoutName]
+	if !ok {
+		diags.Append(diag.NewErrorDiagnostic(
+			"Timeout Does Not Exist",
+			fmt.Sprintf("timeout for %q does not exist", timeoutName),
+		))
+
+		return defaultTimeout, diags
+	}
```
Comment (bflad) [inline datasource/timeouts/timeouts.go:94]: How do you feel about the prior implementation that accepted a default timeout duration inline? I think returning an error diagnostic here could be problematic for provider implementations which elsewhere in the framework do not need to conditionally choose when to append diagnostics to their response diagnostics. I'd expect something like the following to "always work" (configured or not):

```go
timeout, diags := data.Timeouts.Read(ctx) // or Read(ctx, 20*time.Minute)

resp.Diagnostics(diags...)
```

If we wanted to make this nicer for troubleshooting, I'd suggest passing `context.Context` through and logging using `tflog.Info(ctx, timeoutName + " timeout configuration not found, using provided default")`

File: datasource/timeouts/timeouts_test.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+			if diff := cmp.Diff(gotErr, test.expectedDiags, equateErrorMessage); diff != "" {
+				t.Errorf("unexpected err difference: %s", diff)
+			}
+		})
+	}
+}
+
+// equateErrorMessage reports errors to be equal if both are nil
+// or both have the same message.
+var equateErrorMessage = cmp.Comparer(func(x, y error) bool {
```
Comment (bflad) [inline datasource/timeouts/timeouts_test.go:215]: Is this necessary? I haven't come across needing anything special necessary with `go-cmp` comparing `diag.Diagnostics`, but if there is something special it'd be good to figure it out.

File: README.md [hunk 6]
```diff
@@ -36,9 +36,9 @@ Terraform configuration.
 
 #### Block
 
-If your configuration is using a nested block to define timeouts, such as the following:
+The following illustrates nested block syntax for defining timeouts on a resource and a data source.
 
-```terraform
+```hcl
```
Comment (bflad) [inline README.md:41]: Using `terraform` for code blocks should be valid in GitHub: https://github.com/github/linguist/blob/c1c34e5260797b4d598f5ec76f19723bfc5a1894/lib/linguist/languages.yml#L2529-L2544 (and especially helpful should the language ever be split off from generic hcl formatting)

**Review by bflad — APPROVED** [review]
A diagnostics usability question, but otherwise optimistically approving this. ✅

File: datasource/timeouts/timeouts_test.go (hunk 5 above)
Comment (bendbennett) [reply datasource/timeouts/timeouts_test.go:215]: Hmmmm..........have just checked and following the most recent code changes this no longer seems to be required.

**Review by bflad — APPROVED** [review]
Looks good to me 🚀

File: datasource/timeouts/timeouts.go (hunk 4 above)
Comment (bflad) [reply datasource/timeouts/timeouts.go]: I think the super cool thing here is that once we have a custom type for "Go time durations", we can likely remove the `diag.Diagnostics` return completely. 😄

**Pull request merged**